### PR TITLE
Manure management pr

### PR DESCRIPTION
**Pull request by ltp23**

### CODE REVIEW

**Review by varrma — APPROVED** [review]

File: .gitignore [hunk 1]
```diff
@@ -40,3 +40,4 @@ venv/
 .DS_store
 
 .pytest_cache/
+RUFAS/routines/manure/output/graphics/
```
Comment (morrowcj) [inline .gitignore:44]:
```suggestion
RUFAS/routines/manure/output/graphics/

```

File: RUFAS/routines/animal/manure/calf_manure_excretion.py [hunk 2]
```diff
@@ -50,18 +50,34 @@ def manure_calculations(bw, p_feces_excrt, p_urine):
...
+    #         "TSd": total_solids,
+    #         "VSd": 7087.413,  # TODO: Implement with correct equation
+    #         "VSnd": 859.390,  # TODO: Implement with correct equation
+    #         "WIP_frac": WIP_frac,
+    #         "WOP_frac": WOP_frac,
+    #         "p_excrt_manure": p_excrt_manure,
+    #         "p_frac": p_frac,
+    #         "K_manure": 0,
+    #         "CH4_manure": methane_emis
+    #         }
```
Comment (morrowcj) [inline RUFAS/routines/animal/manure/calf_manure_excretion.py]:
```suggestion
```

File: RUFAS/routines/animal/manure/calf_manure_excretion.py [hunk 3]
```diff
@@ -50,18 +50,34 @@ def manure_calculations(bw, p_feces_excrt, p_urine):
...
+            "Mkg": 9.07,  # kg
+            "TSd": 1.09,  # kg
+            "VSd": 0.61,  # kg
+            "VSnd": 0.30,  # kg
+            "WIP_frac": 0.00022,  # dimensionless
+            "WOP_frac": 0.00002,  # dimensionless
+            "p_excrt_manure": 0.00001,  # kg
+            "p_frac": 0.00044,  # dimensionless
+            "K_manure": 0.03,  # kg
+            "CH4_manure": 1.0  # kg
```
Comment (morrowcj) [inline RUFAS/routines/animal/manure/calf_manure_excretion.py]: Can we rename some of these variables? Is it feasible to do so, or would it be too difficult to change them everywhere?

`U`, `TAN_s`, `MN`, `Mkg`, `Tsd`, `Vsd` and `VSnd` are hard to understand. "WIP" and "WOP" might be OK, if they are common acronyms, but I'd consider changing them as well. What about `manure_phosphorus_extract`, `phosphorus_fraction`, `manure_potassium`, and `manure_methane` for the last 4?

File: RUFAS/routines/animal/manure/calf_manure_excretion.py [hunk 4]
```diff
@@ -50,18 +50,34 @@ def manure_calculations(bw, p_feces_excrt, p_urine):
...
+            "TAN_s": 0.03,  # g/L
+            "MN": 0.05,  # kg
+            "Mkg": 9.07,  # kg
+            "TSd": 1.09,  # kg
+            "VSd": 0.61,  # kg
+            "VSnd": 0.30,  # kg
+            "WIP_frac": 0.00022,  # dimensionless
+            "WOP_frac": 0.00002,  # dimensionless
+            "p_excrt_manure": 0.00001,  # kg
+            "p_frac": 0.00044,  # dimensionless
```
Comment (morrowcj) [inline RUFAS/routines/animal/manure/calf_manure_excretion.py]:
```suggestion
            "WIP_frac": 0.00022,  # unitless
            "WOP_frac": 0.00002,  # unitless
            "p_excrt_manure": 0.00001,  # kg
            "p_frac": 0.00044,  # unitless
```

File: RUFAS/routines/animal/manure/dry_cow_manure_excretion.py [hunk 5]
```diff
@@ -82,18 +82,34 @@ def manure_calculations(ration_formulation, feed, bw, milk_prod, p_feces_excrt,
...
+    #         "TSd": total_solids,
+    #         "VSd": 7087.413,  # TODO: Implement with correct equation
+    #         "VSnd": 859.390,  # TODO: Implement with correct equation
+    #         "WIP_frac": WIP_frac,
+    #         "WOP_frac": WOP_frac,
+    #         "p_excrt_manure": p_excrt_manure,
+    #         "p_frac": p_frac,
+    #         "K_manure": K_manure,
+    #         "CH4_manure": methane_emis
+    #         }
```
Comment (morrowcj) [inline RUFAS/routines/animal/manure/dry_cow_manure_excretion.py]:
```suggestion
```

File: RUFAS/routines/animal/manure/dry_cow_manure_excretion.py [hunk 6]
```diff
@@ -82,18 +82,34 @@ def manure_calculations(ration_formulation, feed, bw, milk_prod, p_feces_excrt,
...
+                "TAN_s": 0.14,  # g/L
+                "MN": 0.19,  # kg
+                "Mkg": 32.21,  # kg
+                "TSd": 4.13,  # kg
+                "VSd": 2.49,  # kg
+                "VSnd": 1.0,  # kg
+                "WIP_frac": 0.00046,  # dimensionless
+                "WOP_frac": 0.00005,  # dimensionless
+                "p_excrt_manure": 0.03,  # kg
+                "p_frac": 0.00092,  # dimensionless
```
Comment (morrowcj) [inline RUFAS/routines/animal/manure/dry_cow_manure_excretion.py]:
```suggestion
                "WIP_frac": 0.00046,  # unitless
                "WOP_frac": 0.00005,  # unitless
                "p_excrt_manure": 0.03,  # kg
                "p_frac": 0.00092,  # unitless
```

File: RUFAS/routines/animal/manure/dry_cow_manure_excretion.py [hunk 7]
```diff
@@ -82,18 +82,34 @@ def manure_calculations(ration_formulation, feed, bw, milk_prod, p_feces_excrt,
...
+                "K_manure": 0.12,  # kg
+                "CH4_manure": 1.2  # kg
+                }
```
Comment (morrowcj) [inline RUFAS/routines/animal/manure/dry_cow_manure_excretion.py]: Same comment about variable names.

File: RUFAS/routines/animal/manure/growing_heifer_manure_excretion.py [hunk 8]
```diff
@@ -66,18 +66,35 @@ def manure_calculations(ration_formulation, feed, bw, p_feces_excrt, p_urine):
...
+    #         "TSd": total_solids,
+    #         "VSd": 7087.413,  # TODO: Implement with correct equation
+    #         "VSnd": 859.390,  # TODO: Implement with correct equation
+    #         "WIP_frac": WIP_frac,
+    #         "WOP_frac": WOP_frac,
+    #         "p_excrt_manure": p_excrt_manure,
+    #         "p_frac": p_frac,
+    #         "K_manure": K_manure,
+    #         "CH4_manure": methane_emis
+    #         }
```
Comment (morrowcj) [inline RUFAS/routines/animal/manure/growing_heifer_manure_excretion.py]:
```suggestion
```

File: RUFAS/routines/animal/manure/growing_heifer_manure_excretion.py [hunk 9]
```diff
@@ -66,18 +66,35 @@ def manure_calculations(ration_formulation, feed, bw, p_feces_excrt, p_urine):
...
+            "TSd": 4.04,  # kg
+            "VSd": 2.45,  # kg
+            "VSnd": 1.00,  # kg
+            "WIP_frac": 0.00036,  # dimensionless
+            "WOP_frac": 0.00004,  # dimensionless
+            "p_excrt_manure": 0.02,  # kg
+            "p_frac": 0.00073,  # dimensionless
+            "K_manure": 0.12,  # kg
+            "CH4_manure": 1.20  # kg
             }
```
Comment (morrowcj) [inline RUFAS/routines/animal/manure/growing_heifer_manure_excretion.py:84]: Same variable name comment

File: RUFAS/routines/animal/manure/lactating_cow_manure_excretion.py [hunk 10]
```diff
@@ -121,18 +121,34 @@ def manure_calculations(ration_formulation, feed, bw, days_milk, milk_protein,
...
+    #         "TSd": total_solids,
+    #         "VSd": degradable_volatile_solids,
+    #         "VSnd": nondegradable_volatile_solids,
+    #         "WIP_frac": WIP_frac,
+    #         "WOP_frac": WOP_frac,
+    #         "p_excrt_manure": p_excrt_manure,
+    #         "p_frac": p_frac,
+    #         "K_manure": K_manure,
+    #         "CH4_manure": methane_emis
+    #         }
```
Comment (morrowcj) [inline RUFAS/routines/animal/manure/lactating_cow_manure_excretion.py]:
```suggestion
```

File: RUFAS/routines/animal/manure/lactating_cow_manure_excretion.py [hunk 11]
```diff
@@ -121,18 +121,34 @@ def manure_calculations(ration_formulation, feed, bw, days_milk, milk_protein,
...
+            "TSd": 9.07,  # kg
+            "VSd": 5.01,  # kg
+            "VSnd": 2.70,  # kg
+            "WIP_frac": 0.00073,  # dimensionless
+            "WOP_frac": 0.00007,  # dimensionless
+            "p_excrt_manure": 0.1,  # kg
+            "p_frac": 0.00146,  # dimensionless
+            "K_manure": 0.21,  # kg
+            "CH4_manure": 1.50  # kg
             }
```
Comment (morrowcj) [inline RUFAS/routines/animal/manure/lactating_cow_manure_excretion.py]: Here too.

File: RUFAS/routines/manure/beddings/bedding_classes.py [hunk 12]
```diff
@@ -25,8 +25,8 @@ class BaseBedding(ABC):
     Attributes:
         bedding_mass_per_day: Amount of bedding needed for each animal per day, kg/animal/day.
         bedding_density: Density of the bedding, kg/m^3.
-        bedding_dry_matter_mass: Dry matter mass of the bedding, kg.
-        bedding_cleaned_frac: Fraction of bedding removed from the barn (0.0 - 1.0).
+        bedding_dry_matter_content: Dry matter mass of the bedding, [0.7 - 1.0].
+        bedding_cleaned_frac: Fraction of bedding removed from the barn [0.7 - 1.0].
```
Comment (morrowcj) [inline RUFAS/routines/manure/beddings/bedding_classes.py]: if one is "content" and the other is "frac", they should have different units/ranges, right? I'm guessing content should be "kg" and frac should be [0.7, 1.0].

File: RUFAS/routines/manure/manure/pen_manure.py [hunk 13]
```diff
@@ -0,0 +1,81 @@
...
+    manure_mass: float = 0.0
+    TS: float = 0.0
+    VSd: float = 0.0
+    VSnd: float = 0.0
+    WIP_frac: float = 0.0
+    WOP_frac: float = 0.0
+    P: float = 0.0
+    P_frac: float = 0.0
+    K: float = 0.0
+    CH4: float = 0.0
```
Comment (morrowcj) [inline RUFAS/routines/manure/manure/pen_manure.py]: change attribute names to be more descriptive?

File: RUFAS/routines/manure/manure_management.py [hunk 14]
```diff
@@ -10,10 +10,24 @@
             Sadman Chowdhury, skc86@cornell.edu
 """
 import collections
-from typing import Dict, List, Tuple
+from typing import Dict
+from typing import List
+from typing import Optional
+from typing import Tuple
```
Comment (morrowcj) [inline RUFAS/routines/manure/manure_management.py]: why separate this into multiple lines?

File: RUFAS/routines/manure/manure_management.py [hunk 15]
```diff
@@ -10,10 +10,24 @@
...
 from RUFAS.routines.animal.animal_management import AnimalManagement
+from RUFAS.routines.manure.beddings.bedding_classes import BaseBedding
+from RUFAS.routines.manure.beddings.bedding_classes import BeddingFactory
+from RUFAS.routines.manure.input_handler.manure_management_config_handler import ManureManagementConfigHandler
+from RUFAS.routines.manure.manure_handlers.manure_handler_classes import BaseManureHandler
+from RUFAS.routines.manure.manure_handlers.manure_handler_classes import ManureHandlerFactory
+from RUFAS.routines.manure.manure_separators.manure_separator_classes import BaseManureSeparator
+from RUFAS.routines.manure.manure_separators.manure_separator_classes import ManureSeparatorFactory
+from RUFAS.routines.manure.manure_treatments.manure_treatment_classes import BaseManureTreatment
+from RUFAS.routines.manure.manure_treatments.manure_treatment_classes import ManureTreatmentFactory
```
Comment (morrowcj) [inline RUFAS/routines/manure/manure_management.py]: Same here: why have multiple import statements for the same file?

File: RUFAS/routines/manure_storage/manure_storage.py [hunk 16]
```diff
@@ -353,8 +353,8 @@ def calibrate_water_use(self):
             elif self.handling_system.startswith("automatic_alley_scrapers"):
                 self.water_use_rate = 100
             else:
-                print(self.handling_system, 'is not currently implemented as a handling method. '
-                                            'Setting to flush system')
+                # print(self.handling_system, 'is not currently implemented as a handling method. '
+                #                             'Setting to flush system')
```
Comment (morrowcj) [inline RUFAS/routines/manure_storage/manure_storage.py]:
```suggestion
```

File: RUFAS/routines/manure_storage/manure_storage.py [hunk 17]
```diff
@@ -375,8 +375,8 @@ def calibrate_bedding(self):
                 self.bedding_dry_matter = 0.9
                 self.bedding_washed_perc = 0.8
             else:
-                print(self.bedding, 'is not currently implemented as a bedding type for manure storage. '
-                                    'Setting to organic bedding')
+                # print(self.bedding, 'is not currently implemented as a bedding type for manure storage. '
+                #                     'Setting to organic bedding')
```
Comment (morrowcj) [inline RUFAS/routines/manure_storage/manure_storage.py]:
```suggestion
```

File: RUFAS/routines/manure_storage/manure_storage.py [hunk 18]
```diff
@@ -518,7 +518,7 @@ def calibrate_separator(self):
                 self.TS_DM_effluent_rate = 0.2
 
             else:
-                print(self.separator, 'is not currently implemented as a separator. Setting to sedimentation')
+                # print(self.separator, 'is not currently implemented as a separator. Setting to sedimentation')
```
Comment (morrowcj) [inline RUFAS/routines/manure_storage/manure_storage.py]:
```suggestion
```

File: RUFAS/routines/manure/units/units.py [hunk 19]
```diff
@@ -0,0 +1,31 @@
...
+    N = 'kg'
+    manure_mass = 'kg'
+    TS = 'kg'
+    VSd = 'kg'
+    VSnd = 'kg'
+    VS_total = 'kg'
+    P = 'kg'
+    K = 'kg'
+    CH4 = 'g_per_day'
+
```
Comment (morrowcj) [inline RUFAS/routines/manure/units/units.py:29]: attribute names again. I'll stop mentioning this. Just assume that I think they should have better names throughout the PR.

File: input/animal_management.json [hunk 20]
```diff
@@ -16,6 +16,7 @@
         "fields": {
         },
         "animal": "animal_management_animal.json",
-        "feed": "panke_buisse_feed.json"
+        "feed": "panke_buisse_feed.json",
+        "manure": "manure_management.json"
     }
 }
```
Comment (morrowcj) [inline input/animal_management.json:23]:
```suggestion
}

```

File: tests/test_manure_module/conftest.py [hunk 21]
```diff
@@ -0,0 +1,122 @@
...
+    mock_pen.animals_in_pen = [mock_cow] * 5
+    mock_pen.classes_in_pen = {'Cow'}
+    mock_pen.animal_combination = Pen.AnimalCombination.LAC_COW
+    mock_pen.housing_type = 'free stall'
+    mock_pen.bedding_type = 'sawdust'
+    mock_pen.manure_handling = 'flush system'
+    mock_pen.manure_separator = 'rotary screen'
+    mock_pen.manure_storage = 'slurry storage outdoor'
+    mock_pen.manure = generate_animal_manure(10.0)
+    return mock_pen
```
Comment (morrowcj) [inline tests/test_manure_module/conftest.py]: maybe rename the file to "manure_test_fixtures.py"?

File: tests/test_manure_module/test_manure.py [hunk 22]
```diff
@@ -0,0 +1,832 @@
+# from typing import Callable
```
Comment (morrowcj) [inline tests/test_manure_module/test_manure.py]: Why is this entire file commented out? If `manure.py` no longer exists, delete this.

File: tests/test_manure_module/test_manure_management_pen.py [hunk 23]
```diff
@@ -0,0 +1,115 @@
...
+from RUFAS.routines.manure.manure.pen_manure import PenManure
+from RUFAS.routines.manure.pen.manure_management_pen import ManureManagementPen
+
+
+def test_manure_management_pen_init(mocker: MockerFixture,
+                                    mock_pen: Pen,
+                                    mock_cow: Cow,
+                                    generate_animal_manure: Callable[[float], Dict[str, float]]) -> None:
+    """Unit test for function __init__ in file manure_management_pen.py"""
+
```
Comment (morrowcj) [inline tests/test_manure_module/test_manure_management_pen.py]: Be more descriptive about what your tests are doing. 
```suggestion
    """Check that manure management is properly initialized, with correct values"""
```

I'd recommending all your test doscstrings to be more descriptive, but I'll only comment once.

File: tests/test_manure_module/test_manure_management_pen.py [hunk 24]
```diff
@@ -0,0 +1,115 @@
...
+    # Arrange
+    mock_pen.housing_type = housing_type
+    mock_pen.animals_in_pen = [mock_cow] if has_cows else [mock_calf]
+    mock_pen.classes_in_pen = {'Cow'} if has_cows else {'Calf'}
+
+    # Act
+    mm_pen = ManureManagementPen(mock_pen)
+
+    # Assert
+    assert mm_pen.barn_area == approx(expected_area)
```
Comment (morrowcj) [inline tests/test_manure_module/test_manure_management_pen.py]: Aren't there more components of `manure_management_pen.py` that need testing? What about the class properties?

File: tests/test_manure_module/test_pen_manure.py [hunk 25]
```diff
@@ -0,0 +1,64 @@
...
+    # assert manure.K == approx(8.0 * constants.GRAMS_TO_KG)
+
+    assert manure.manure_mass == approx(10.0)
+    assert manure.CH4 == approx(10.0)
+
+    # The remaining attributes should be set to the default value of 0.
+    assert manure.TS == approx(0.0)
+    assert manure.WIP_frac == approx(0.0)
+    assert manure.WOP_frac == approx(0.0)
+    assert manure.P_frac == approx(0.0)
```
Comment (morrowcj) [inline tests/test_manure_module/test_pen_manure.py]: no test for `get_instance`?

**Review by morrowcj — APPROVED** [review]
To the extent that I understand it, this PR looks pretty good. I do think there is room for further cleaning and refactoring in the older files, but the files that are completely new look really good. 

In general, I would recommend trying to shift towards more descriptive names and docstrings. 

Also I think the testing could be a bit more extensive. I'd prefer to see more smaller tests, broken into chunks that test each individual piece (e.g., does member function `x` work properly?) in addition to the larger integration-type tests that check the larger, overall functionality. 

Generally, I think this is good, though. I'm going to go ahead and approve because none of my suggested changes are too substantive.

For reference, it took me about 3.5 hours to review this PR.

**Comment by morrowcj** [comment]
@ltp23, it seems like a lot of the constants in your various files would be applicable to the whole model (`general_constants.py`) that should maybe be higher up in the tree.

**Pull request closed without merging**